Adding game controlers as input method

nuess0r opened the pull request:
I had a Bluetooth joypad lying around and though that it would be handy to be able to remote control the projectM frontend.

As SDL provides everything needed to handle game controller it was straight forward to add it. I didn't add SDL controller class or so, just extended the SDL window class. Feel free to refactor that if you think this is messy.

Tested on GNU/Linux, Debian Sid, Gnome3

Code comes mainly from this tutorial: https://lazyfoo.net/tutorials/SDL/19_gamepads_and_joysticks/index.php

nuess0r committed:
Adding game controller support

Using the SDL2 library to use gamecontrollers/joypads as user input method.

nuess0r committed:
Add description of GUI controls to README.md

revmischa commented:
Nice!

kblaschke commented:
That looks quite good, and should surely work on all platforms!

Lemme fix the Windows build issue tomorrow, as the build check workflow is broken (after GitHub Actions Windows runners being broken and then fixed again). You may have to rebase on latest master after the fix is pushed to make the build check work, as the workflow file is used from the branch.

kblaschke commented:
The Windows build check is now fixed.

Please rebase onto current master and force-push your branch, then the checks here will probably all be green.

kblaschke commented on src/SDLRenderingWindow.cpp:
```diff
@@ -461,3 +461,21 @@ void SDLRenderingWindow::OnConfigurationPropertyRemoved(const std::string& key)
+        poco_debug(_logger, "No joysticks connected");
+    }
+
+    //For simplicity, we’ll only be setting up and tracking a single
```
If you want, you could also add a setting and an option in the settings dialog to let the user select a specific controller. Then select the stored one automatically on startup, e.g. via [its name](https://wiki.libsdl.org/SDL2/SDL_GameControllerNameForIndex) or [device path](https://wiki.libsdl.org/SDL2/SDL_GameControllerPathForIndex) (which should be map to the port it is connected to).

If you're not into Dear ImGui, I can also add this stuff at a later time.

nuess0r replied:
Didn't knew about Dear ImGui before I had a look at this project. Adding features to the GUI is a bit over my head at the moment (my available time is scrace). Actually I'm not even into C++ but this doesn't hinder me to tinker around :-)

kblaschke replied:
Sure, no problem! I've got the same time issue, but I'll try my best to spend a bit of it on projectM every week 😀 
I'll review the changes this weekend more thoroughly, but at a glance it all looks fine, so expect it to be merged soon!

Thanks again for the contribution!

nuess0r committed:
Implementing review: Move _controller pointer to SDLRenderingWindow class

nuess0r committed:
Move debug log level example further down, so it actually works

When you uncommented the existing debug log level example it had no effect
because the default level (informative) is set further down in the config
file and therefore took priority.

kblaschke commented:
Sorry for the long stall - I've not forgotten about this, just currently not actively working on the frontend and focusing on the core library. Will review and eventually merge this once the libprojectM 4.2 release is out!

kblaschke committed:
Use C++17 to build if we find POCO 1.14 or higher.

kblaschke commented:
Rebased your branch onto PR #97 to make the build checks work again - please update your local working branch accordingly to avoid conflicts if you want to add something.

kblaschke commented on README.md:
```diff
@@ -7,6 +7,28 @@ It will listen to audio input and produce mesmerizing visuals. Some commands are
+
+| Command | Keyboard | Mouse | Game Controller |
+| --- | --- | --- | --- |
+| Add a new waveform | | `ctrl` + Left |  |
+| Clear all custom waveforms | | Middle |  |
```
This functionality was not reimplemented when rewriting the renderer, as it doesn't really fit the structure well and was quite cumbersome to use anyway. No need to add controls for something that doesn't work.
```suggestion
```

kblaschke commented on src/resources/projectMSDL.properties.in:
```diff
@@ -145,9 +140,13 @@ logging.channels.async.class = AsyncChannel
-logging.loggers.root.level = information
 logging.loggers.root.channel = async
 
+# Set log level to debug for all components
+#logging.loggers.root.level = debug
+logging.loggers.root.level = information
+
+
```
Should remove this addition here before merging, was probably a leftover from testing.
```suggestion
```

nuess0r committed:
Update src/resources/projectMSDL.properties.in

As suggested by @kblaschke

Co-authored-by: Kai Blaschke <kai.blaschke@kb-dev.net>

nuess0r committed:
As suggested by @kblaschke

kblaschke commented:
Sorry for the long wait.

I'm (almost) ready to pick up on this again, so I'd then go ahead and rebase your branch on top of the lastest master and other incoming changes, shuffle a few things around and add a settings tab to the UI for configuring the game controller.

I can do this in your fork and this PR if you're okay with that, or I can copy it over to my fork and open a new PR.
